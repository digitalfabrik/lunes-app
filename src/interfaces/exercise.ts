export interface IVocabularyOverviewScreen {
  navigation: any
  route: any
}

export interface IVocabularyTrainerScreen {
  navigation: any
  route: any
}

export interface IAlternativeWordProps {
  alt_word: string
  article: string
}

export interface IDocumentProps {
  id: number
  word: string
  article: string
  document_image: Array<{ id: number; image: string }>
  audio: string
  alternatives: IAlternativeWordProps[]
}

export interface IVocabularyOverviewListItemProps {
  id: number
  word: string
  article: string
  image: string
  audio: string
}

export interface IConfirmationModalProps {
  visible: boolean
  setIsModalVisible: Function
  navigation: any
  extraParams: any
}

export interface IAnswerSectionProps {
  tryLater: () => void
  currentDocumentNumber: number
  setCurrentDocumentNumber: Function
  documents: IDocumentProps[]
  finishExercise: Function
  trainingSet: string
  disciplineTitle: string
}

export interface IFeedbackProps {
  secondAttempt: boolean
  result: string
  document?: IDocumentProps
  input: string
}

export interface IActionsProps {
  tryLater: () => void
  giveUp: () => void
  result: string
  checkEntry: () => void
  getNextWord: () => void
  input: string
  isFinished: boolean
  secondAttempt: boolean
}
