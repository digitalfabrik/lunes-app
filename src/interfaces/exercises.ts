export interface IExercisesScreenProps {
  route: any
  navigation: any
}
