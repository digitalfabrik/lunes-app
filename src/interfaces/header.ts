export interface IHeaderProps {
  top: number | undefined
}
