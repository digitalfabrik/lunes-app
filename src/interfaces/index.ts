import { ReactElement } from 'react'

export interface ILoadingProps {
  children: ReactElement
  isLoading: boolean
}

export interface IButtonProps {
  onPress: () => void
  disabled?: boolean
  children: ReactElement
  theme?: string
  testID?: string
}
export interface IPopoverProps {
  setIsPopoverVisible: (param: boolean) => void
  isVisible: boolean
  children: ReactElement
}
