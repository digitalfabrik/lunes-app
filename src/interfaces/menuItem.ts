import { ReactElement } from 'react'

export interface IMenuItemProps {
  selected: boolean
  onPress: () => void
  icon: string
  title: string
  children: ReactElement
}
