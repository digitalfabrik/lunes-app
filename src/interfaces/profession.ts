export interface IProfessionScreenProps {
  navigation: any
}

export interface IProfessionsProps {
  id: number
  title: string
  description: string
  icon?: string
  total_training_sets: number
}
