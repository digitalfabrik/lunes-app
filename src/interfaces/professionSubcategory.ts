export interface IProfessionSubcategoryScreenProps {
  route: any
  navigation: any
}

export interface IProfessionSubcategoryProps {
  id: number
  title: string
  description: string
  icon: string
  total_documents: number
}
