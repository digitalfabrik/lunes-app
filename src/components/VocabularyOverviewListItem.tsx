import React from 'react'
import { IVocabularyOverviewListItemProps } from '../interfaces/exercise'
import { View, Text, Image, TouchableOpacity, Platform, StyleSheet } from 'react-native'
import { COLORS } from '../constants/colors'
import { ARTICLES } from '../constants/data'
import Tts from 'react-native-tts'
import { getArticleColor, capitalizeFirstLetter } from '../utils/helpers'
import SoundPlayer from 'react-native-sound-player'
import { VolumeUp } from '../../assets/images'
import { widthPercentageToDP as wp } from 'react-native-responsive-screen'

export const styles = StyleSheet.create({
  wrapper: {
    paddingHorizontal: 16
  },
  container: {
    paddingVertical: 17,
    paddingHorizontal: 16,
    marginBottom: 8,
    flexDirection: 'row',
    justifyContent: 'space-between',
    alignItems: 'center',
    width: '100%',
    backgroundColor: COLORS.white,
    borderColor: COLORS.lunesBlackUltralight,
    borderWidth: 1,
    borderStyle: 'solid',
    borderRadius: 2
  },
  item: {
    flexDirection: 'row',
    alignItems: 'center'
  },
  image: {
    marginRight: 15,
    width: wp('15%'),
    height: wp('15%'),
    borderRadius: 50
  },
  title: {
    fontSize: wp('3.5%'),
    fontWeight: 'normal',
    borderRadius: 10,
    marginBottom: 6,
    color: COLORS.lunesGreyDark,
    fontFamily: 'SourceSansPro-Regular',
    alignSelf: 'flex-start',
    width: wp('10%'),
    overflow: 'hidden',
    height: wp('5%'),
    textAlign: 'center',
    lineHeight: 18
  },
  description: {
    fontSize: wp('4%'),
    fontWeight: 'normal',
    color: COLORS.lunesGreyMedium,
    fontFamily: 'SourceSansPro-Regular',
    marginLeft: 8
  },
  speaker: {
    borderRadius: 50,
    justifyContent: 'center',
    alignItems: 'center'
  },
  shadow: {
    shadowColor: COLORS.shadow,
    shadowOffset: { width: 0, height: 5 },
    shadowRadius: 3,
    shadowOpacity: 10
  }
})

// German language
Tts.setDefaultLanguage('de-DE')

const VocabularyOverviewListItem = ({ image, article, word, audio }: IVocabularyOverviewListItemProps) => {
  const [active, setActive] = React.useState(false)
  React.useEffect(() => {
    const _onSoundPlayerFinishPlaying = SoundPlayer.addEventListener('FinishedPlaying', () => setActive(false))

    const _onTtsFinishPlaying = Tts.addEventListener('tts-finish', () => setActive(false))

    return () => {
      _onSoundPlayerFinishPlaying.remove()
      _onTtsFinishPlaying.remove()
    }
  }, [])

  const handleSpeakerClick = () => {
    setActive(true)

    // Don't use soundplayer for IOS, since IOS doesn't support .ogg files
    if (audio && Platform.OS !== 'ios') {
      // audio from API
      SoundPlayer.playUrl(audio)
    } else {
      Tts.speak(`${article} ${word}`, {
        androidParams: {
          KEY_PARAM_PAN: -1,
          KEY_PARAM_VOLUME: 0.5,
          KEY_PARAM_STREAM: 'STREAM_MUSIC'
        }
      })
    }
  }

  const volumeIconColor = active ? COLORS.lunesRed : COLORS.lunesRedDark

  const volumeIconStyle = [styles.speaker, !active && styles.shadow]

  return (
    <View style={styles.wrapper}>
      <View style={styles.container}>
        <View style={styles.item}>
          <Image
            source={{
              uri: image
            }}
            width={24}
            height={24}
            style={styles.image}
          />
          <View>
            <Text testID='article' style={[styles.title, { backgroundColor: getArticleColor(article) }]}>
              {article?.toLowerCase() === ARTICLES.diePlural ? 'Die' : capitalizeFirstLetter(article)}
            </Text>
            <Text testID='word' style={styles.description}>
              {word}
            </Text>
          </View>
        </View>
        <TouchableOpacity testID='volume-button' style={volumeIconStyle} onPress={() => handleSpeakerClick()}>
          <VolumeUp fill={volumeIconColor} />
        </TouchableOpacity>
      </View>
    </View>
  )
}

export default VocabularyOverviewListItem
