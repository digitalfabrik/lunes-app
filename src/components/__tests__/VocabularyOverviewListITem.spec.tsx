import 'react-native'
import React from 'react'
import VocabularyOverviewListItem from '../VocabularyOverviewListItem'
import { shallow } from 'enzyme'
import toJson from 'enzyme-to-json'
import { IVocabularyOverviewListItemProps } from '../../interfaces/exercise'

describe('Components', () => {
  describe('VocabularyOverviewListItem', () => {
    const defaultListItemProps: IVocabularyOverviewListItemProps = {
      article: '',
      audio: '',
      id: 0,
      image: '',
      word: ''
    }

    it('renders correctly across screens', () => {
      const component = shallow(<VocabularyOverviewListItem {...defaultListItemProps} />)
      expect(toJson(component)).toMatchSnapshot()
    })

    it('should display image passed to it', () => {
      const listItemProps: IVocabularyOverviewListItemProps = {
        ...defaultListItemProps,
        image: 'https://lunes.tuerantuer.org/media/images/Winkelmesser.jpeg'
      }

      const component = shallow(<VocabularyOverviewListItem {...listItemProps} />)
      expect(component.find('Image').prop('source')).toHaveProperty('uri', listItemProps.image)
    })

    it('should display article passed to it', () => {
      const listItemProps: IVocabularyOverviewListItemProps = {
        ...defaultListItemProps,
        article: 'article'
      }
      const article = 'Article'

      const component = shallow(<VocabularyOverviewListItem {...listItemProps} />)
      expect(component.find('[testID="article"]').props().children).toBe(article)
    })

    it('should display word passed to it', () => {
      const listItemProps: IVocabularyOverviewListItemProps = {
        ...defaultListItemProps,
        word: 'word'
      }
      const word = 'word'

      const component = shallow(<VocabularyOverviewListItem {...listItemProps} />)
      expect(component.find('[testID="word"]').props().children).toBe(word)
    })

    it('should render volume button', () => {
      const listItemProps: IVocabularyOverviewListItemProps = {
        ...defaultListItemProps
      }

      const component = shallow(<VocabularyOverviewListItem {...listItemProps} />)
      expect(component.find('[testID="volume-button"]')).toHaveLength(1)
    })
  })
})
