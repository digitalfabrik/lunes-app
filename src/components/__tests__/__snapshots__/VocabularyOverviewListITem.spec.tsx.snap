// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Components VocabularyOverviewListItem renders correctly across screens 1`] = `
<View
  style={
    Object {
      "paddingHorizontal": 16,
    }
  }
>
  <View
    style={
      Object {
        "alignItems": "center",
        "backgroundColor": "#ffffff",
        "borderColor": "#e0e4ed",
        "borderRadius": 2,
        "borderStyle": "solid",
        "borderWidth": 1,
        "flexDirection": "row",
        "justifyContent": "space-between",
        "marginBottom": 8,
        "paddingHorizontal": 16,
        "paddingVertical": 17,
        "width": "100%",
      }
    }
  >
    <View
      style={
        Object {
          "alignItems": "center",
          "flexDirection": "row",
        }
      }
    >
      <Image
        height={24}
        source={
          Object {
            "uri": "",
          }
        }
        style={
          Object {
            "borderRadius": 50,
            "height": 112.5,
            "marginRight": 15,
            "width": 112.5,
          }
        }
        width={24}
      />
      <View>
        <Text
          style={
            Array [
              Object {
                "alignSelf": "flex-start",
                "borderRadius": 10,
                "color": "#3a4049",
                "fontFamily": "SourceSansPro-Regular",
                "fontSize": 26.5,
                "fontWeight": "normal",
                "height": 37.5,
                "lineHeight": 18,
                "marginBottom": 6,
                "overflow": "hidden",
                "textAlign": "center",
                "width": 75,
              },
              Object {
                "backgroundColor": "#8cc8f3",
              },
            ]
          }
          testID="article"
        />
        <Text
          style={
            Object {
              "color": "#5b616a",
              "fontFamily": "SourceSansPro-Regular",
              "fontSize": 30,
              "fontWeight": "normal",
              "marginLeft": 8,
            }
          }
          testID="word"
        />
      </View>
    </View>
    <ForwardRef
      onPress={[Function]}
      style={
        Array [
          Object {
            "alignItems": "center",
            "borderRadius": 50,
            "justifyContent": "center",
          },
          Object {
            "shadowColor": "rgba(0, 0, 0, 0.6)",
            "shadowOffset": Object {
              "height": 5,
              "width": 0,
            },
            "shadowOpacity": 10,
            "shadowRadius": 3,
          },
        ]
      }
      testID="volume-button"
    >
      <[object Object]
        fill="#F1635F"
      />
    </ForwardRef>
  </View>
</View>
`;
